BSP/Tests/ADC.c

```diff
@@ -0,0 +1,390 @@
+    * Input: GPIO_PIN_4 on GPIOA
+    * LED3 (on board) turns ON when error occurs
+
+    * Main.h is easy to change so that you can run it on different pins 
+    * Using SysTick - be Careful with timer
+    * Have to have certain stuff in certain directories to build properly
+    * Runs on our device
+    * Been tested (by STM?) w/ NUCLEO-L432KC board
+
+    * How to use?
+        - Rebuild all files and flash into memory
+        - Run the example
+        - watch aADCxConvertedData varaible in debugger live watch
+
+    * From Main.c, basically warnings so you dont mess up 
+        /* STM32L4xx HAL library initialization:
+            - Configure the Flash prefetch
+            - Systick timer is configured by default as source of time base, but user 
+                can eventually implement his proper time base source (a general purpose 
+                timer for example or other time source), keeping in mind that Time base 
+                duration should be kept 1ms since PPP_TIMEOUT_VALUEs are defined and 
+                handled in milliseconds basis.
+            - Set NVIC Group Priority to 4
+            - Low Level Initialization
+            */ 
+    */
+
+
+// Main.h stuff:
+/**
+  ******************************************************************************
+  * @file    ADC/ADC_DMA_Transfer/Inc/main.h
+  * @author  MCD Application Team
+  * @brief   Header for main.c module
+  ******************************************************************************
+  * @attention
+  *
+  * Copyright (c) 2017 STMicroelectronics.
+  * All rights reserved.
+  *
+  * This software is licensed under terms that can be found in the LICENSE file
+  * in the root directory of this software component.
+  * If no LICENSE file comes with this software, it is provided AS-IS.
+  *
+  ******************************************************************************
+  */
+
+/* Define to prevent recursive inclusion -------------------------------------*/
+#ifndef __MAIN_H
+#define __MAIN_H
+
+/* Includes ------------------------------------------------------------------*/
+#include "stm32l4xx_hal.h"
+#include "stm32l4xx_nucleo_32.h"
+
+/* Exported types ------------------------------------------------------------*/
+/* Exported constants --------------------------------------------------------*/
+/* User can use this section to tailor ADCx instance used and associated
+   resources */
+   
+/* Definition for ADCx clock resources */
+#define ADCx                            ADC1                        // dont really know whats going on here 
+#define ADCx_CLK_ENABLE()               __HAL_RCC_ADC_CLK_ENABLE()
+#define ADCx_CHANNEL_GPIO_CLK_ENABLE()  __HAL_RCC_GPIOA_CLK_ENABLE()
+
+#define DMAx_CHANNELx_CLK_ENABLE()      __HAL_RCC_DMA1_CLK_ENABLE()
+
+#define ADCx_FORCE_RESET()              __HAL_RCC_ADC_FORCE_RESET()
+#define ADCx_RELEASE_RESET()            __HAL_RCC_ADC_RELEASE_RESET()
```

Comment:
You'll understand this a bit more once you do your ADC lab in 319 lol. TLDR, you're just turning on multiple clocks within that pin in order to sample the analog sample that's coming in. The way it works is that you're taking some signal (let's say a sine wave for the sake of visualization) and you're sampling it at a specific frequency which will tell your software what the wave looks like. Lmk if you have more specific questions since that was a bit of a poor explanation. If you want to do research I'd straight up look up "STM32 HAL ADC init" or read the STM32L4 HAL documentation pdf.